src/mito2/src/compaction/twcs.rs

```diff
@@ -0,0 +1,831 @@
+        windows.entry(time_window).or_default().push(file.clone());
+    }
+    windows
+}
+
+/// Finds the latest active writing window among all files.
+/// Returns `None` when there are no files or all files are corrupted.
+fn find_latest_window_in_seconds<'a>(
+    files: impl Iterator<Item = &'a FileHandle>,
+    time_window_size: i64,
+) -> Option<i64> {
+    let mut latest_timestamp = None;
+    for f in files {
+        let (_, end) = f.time_range();
+        if let Some(latest) = latest_timestamp && end > latest {
+            latest_timestamp = Some(end);
+        } else {
+            latest_timestamp = Some(end);
+        }
+    }
+    latest_timestamp
+        .and_then(|ts| ts.convert_to_ceil(TimeUnit::Second))
+        .and_then(|ts| ts.value().align_to_ceil_by_bucket(time_window_size))
+}
+
+pub(crate) struct TwcsCompactionTask {
+    pub region_id: RegionId,
+    pub schema: RegionMetadataRef,
+    pub sst_layer: AccessLayerRef,
+    pub outputs: Vec<CompactionOutput>,
+    pub expired_ssts: Vec<FileHandle>,
+    pub sst_write_buffer_size: ReadableSize,
+    pub compaction_time_window: Option<i64>,
+    pub file_purger: FilePurgerRef,
+    /// Request sender to notify the worker.
+    pub(crate) request_sender: mpsc::Sender<WorkerRequest>,
+    /// Sender that are used to notify waiters waiting for pending compaction tasks.
+    pub sender: Option<Sender<error::Result<Output>>>,
+}
+
+impl Debug for TwcsCompactionTask {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TwcsCompactionTask").finish()
+    }
+}
+
+impl Drop for TwcsCompactionTask {
+    fn drop(&mut self) {
+        self.mark_files_compacting(false)
+    }
+}
+
+impl TwcsCompactionTask {
+    fn mark_files_compacting(&self, compacting: bool) {
+        self.outputs
+            .iter()
+            .flat_map(|o| o.inputs.iter())
+            .for_each(|f| f.set_compacting(compacting))
+    }
+
+    /// Merges all SST files.
+    /// Returns `(output files, input files)`.
+    async fn merge_ssts(&mut self) -> error::Result<(Vec<FileMeta>, Vec<FileMeta>)> {
+        let mut futs = Vec::with_capacity(self.outputs.len());
+        let mut compacted_inputs = vec![];
+        let region_id = self.region_id;
+        for output in self.outputs.drain(..) {
+            let schema = self.schema.clone();
+            let sst_layer = self.sst_layer.clone();
+            let sst_write_buffer_size = self.sst_write_buffer_size;
+            compacted_inputs.extend(output.inputs.iter().map(FileHandle::meta));
+
+            info!(
+                "Compaction output [{}]-> {}",
+                output
+                    .inputs
+                    .iter()
+                    .map(|f| f.file_id().to_string())
+                    .collect::<Vec<_>>()
+                    .join(","),
+                output.output_file_id
+            );
+
+            // TODO(hl): Maybe spawn to runtime to exploit in-job parallelism.
```

Comment:
Use the background runtime.

Reply:
> Use the background runtime.

We have to do a bench to determine whether spawn works better.